### PR TITLE
Replace Encoder Group with Encoder Pair

### PR DESCRIPTION
**Pull request by carturn**
Encoder Group seems to have limited use cases and limitations due to it using N-encoders.
Therefore, I am proposing replacing it with a pair of encoders, which I think covers most of our uses for multiple encoders together. I think this will simplify implementation of encoder based chassis driving systems.

### CODE REVIEW

**Comment by carturn** [comment]
@ave4224 would this help with the encoder based chassis move you were working on?

File: custom/sensors/EncoderPair.java [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+		}
+		catch (Exception e) {
+			LogKitten.ex(e);
+			return 0;
+		}
+	}
+	
+	@Override
+	public boolean getDirection() {
+		return (getRate() > 0);
```
Comment (leijurv) [inline custom/sensors/EncoderPair.java]: Parens.

File: custom/sensors/EncoderPair.java [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+	}
+	
+	@Override
+	public boolean getDirectionSafely() throws InvalidSensorException {
```
Comment (leijurv) [inline custom/sensors/EncoderPair.java]: Parens

File: custom/sensors/EncoderPair.java [hunk 3]
```diff
@@ -0,0 +1,289 @@
...
+	public double getRateSafely() throws InvalidSensorException {
+		return (encoders[0].getRateSafely() + encoders[1].getRateSafely()) / 2.0;
+	}
+	
+	@Override
+	public double getRate() {
+		try {
+			return getRateSafely();
+		}
+		catch (Exception e) {
```
Comment (leijurv) [inline custom/sensors/EncoderPair.java]: Catch the specific checked exception not a general exception. Or use | syntax

File: custom/sensors/EncoderPair.java [hunk 4]
```diff
@@ -0,0 +1,289 @@
...
+	public int getSafely() throws InvalidSensorException {
+		return (int) Math.round(((double) (encoders[0].getSafely() - offset[0]) + (double) (encoders[1].getSafely() - offset[1])) / 2.0);
+	}
+	
+	@Override
+	public int get() {
+		try {
+			return getSafely();
+		}
+		catch (Exception e) {
```
Comment (leijurv) [inline custom/sensors/EncoderPair.java]: Same here

File: custom/sensors/EncoderPair.java [hunk 5]
```diff
@@ -0,0 +1,289 @@
...
+public class EncoderPair implements CustomEncoder {
+	private final CustomEncoder[] encoders;
+	private final int[] offset; // Do not reset encoders, just store an offset value
+	private PIDSourceType pidSource;
+	private boolean reverseDirection;
+	private double distancePerPulse;
+	private final double distanceTolerance;
+	private final double rateTolerance;
+	private static final double DEFAULT_DISTANCE_TOLERANCE = 10;
+	private static final double DEFAULT_RATE_TOLERANCE = 10;
```
Comment (leijurv) [inline custom/sensors/EncoderPair.java]: Protected?

File: custom/sensors/EncoderPair.java [hunk 6]
```diff
@@ -0,0 +1,289 @@
...
+		this.distancePerPulse = distancePerPulse;
+		for (CustomEncoder encoder : encoders) {
+			encoder.setDistancePerPulse(distancePerPulse);
+		}
+	}
+	
+	@Override
+	public void reset() {
+		for (int i = 0; i < 2; i++) {
+			offset[i] = encoders[i].get();
```
Comment (leijurv) [inline custom/sensors/EncoderPair.java]: For loop?

File: custom/sensors/EncoderPair.java [hunk 7]
```diff
@@ -0,0 +1,289 @@
...
+	
+	@Override
+	public double getDistancePerPulse() {
+		return distancePerPulse;
+	}
+	
+	@Override
+	public void setDistancePerPulse(double distancePerPulse) {
+		this.distancePerPulse = distancePerPulse;
+		for (CustomEncoder encoder : encoders) {
```
Comment (leijurv) [inline custom/sensors/EncoderPair.java]: Enhanced for each loop?

File: custom/sensors/EncoderPair.java [hunk 8]
```diff
@@ -0,0 +1,289 @@
...
+	 * This respects the original inversion state of each encoder when constructed,
+	 * and will only invert encoders if this.getReverseDirection() != the input.
+	 *
+	 * @param reverseDirection
+	 *        The state of inversion, true is inverted.
+	 */
+	@Override
+	public void setReverseDirection(boolean reverseDirection) {
+		if (getReverseDirection() != reverseDirection) {
+			for (CustomEncoder encoder : encoders) {
```
Comment (leijurv) [inline custom/sensors/EncoderPair.java]: Enhanced for each loop?

File: custom/sensors/EncoderPair.java [hunk 9]
```diff
@@ -0,0 +1,289 @@
...
+	/**
+	 * Sets the direction inversion of all encoder substituents.
+	 * This respects the original inversion state of each encoder when constructed,
+	 * and will only invert encoders if this.getReverseDirection() != the input.
+	 *
+	 * @param reverseDirection
+	 *        The state of inversion, true is inverted.
+	 */
+	@Override
+	public void setReverseDirection(boolean reverseDirection) {
```
Comment (leijurv) [inline custom/sensors/EncoderPair.java]: Unclear naming, local variable hides a field, using this. For differentiation is confusing

File: custom/sensors/EncoderPair.java [hunk 10]
```diff
@@ -0,0 +1,289 @@
...
+		}
+		catch (Exception e) {
+			LogKitten.ex(e);
+			return 0;
+		}
+	}
+	
+	@Override
+	public double getDistanceSafely() throws InvalidSensorException {
+		return ((encoders[0].getDistanceSafely() - offset[0] * encoders[0].getDistancePerPulse()) + (encoders[1].getDistanceSafely() - offset[1] * encoders[1].getDistancePerPulse())) / 2.0;
```
Comment (leijurv) [inline custom/sensors/EncoderPair.java]: No for loop here?

**Comment by carturn** [comment]
Is this something we actually want to do in general?
I would prefer not to waste time cleaning this up if we are not going to use it.

**Comment by leijurv** [comment]
Aren't there various encoder pairs on the chassis?

File: custom/sensors/EncoderPair.java (hunk 9 above)
Comment (carturn) [inline custom/sensors/EncoderPair.java]: This has generally been our convention for similar setters; I am not sure this function is different enough to warrant a change.

File: custom/sensors/EncoderPair.java (hunk 8 above)
Comment (carturn) [inline custom/sensors/EncoderPair.java]: Is there an alternative that would work better here?

**Comment by carturn** [comment]
This is partially tested; should we test the PID capabilities of this before merging? I am inclined to think that given that the getDistance and getDifference functions work, the PID will work as well.

**Pull request merged**